vertical-pod-autoscaler/enhancements/3913-customized-recomender-vpa.md

```diff
@@ -0,0 +1,272 @@
+behaviors for workloads and applying different algorithms to improve resource utilization 
+(CPU and Memory) predictions can significantly reduce over-provisioning and OOM kills in VPA.
+
+### Goals
+
+- Allow the VPA object to specify a customized recommender to use. 
+- Allow the VPA object to use the default recommender when no recommender is specified.
+
+### Non-Goals
+
+- We assume no pod uses two recommenders at the same time.
+- We do not resolve conflicts between recommenders.
+
+## Proposal
+
+### User Stories
+
+#### Story 1 
+
+- Containers with Cyclic Patterns in Resource Usage
+
+Containers used in monitoring may receive load periodically to process but need to be long-running 
+to listen to incoming traffic. Thus, these containers usually exhibit cyclic patterns, alternating 
+between usage spikes and idling. Resizing containers according to usage observed in the previous 
+time window may always lead to under-provision for a short period when the load spikes just arrive. 
+The problem will even lead to frequent pod evictions when the pod was resized according to the idling 
+period and the host resource has been taken by other pods.
+
+#### Story 2 
+
+- Containers with Different but Recurrent Behaviors in Resource Usage
+
+Containers running spark/deep learning training workloads are known to show recurring and repeating 
+patterns in resource usage. Prior research has shown that different but recurrent behaviors occur 
+for different containerized tasks, such as Spark or deep learning training. These common patterns can 
+be represented by phases, which display similar resource usage of computational resources over time. 
+There are common sequences of patterns for different executions of the workload and they can be used 
+to proactively predict future resource usage more accurately. The default recommender in the current 
+VPA adopts a reactive approach so a more proactive recommender is needed for these types of workload.
+
+### Implementation Details
+
+The following describes the details of implementing a first-citizen approach to support the customized 
+recommender. Namely, a dedicated field `recommenderName` is added to the VPA crd definition in 
+`deploy/vpa-v1.crd.yaml`.
+
+```yaml
+validation:
+ # openAPIV3Schema is the schema for validating custom objects.
+ openAPIV3Schema:
+   type: object
+   properties:
+     spec:
+       type: object
+       required: []
+       properties:
+         recommenderName:
+           type: string
+         targetRef:
+           type: object
+         updatePolicy:
+           type: object
+```
+
+Correspondingly, the `VerticalPodAutoscalerSpec` in `pkg/apis/autoscaling.k8s.io/v1/types.go` 
+should be updated to include the `recommenderName` field.
+
+```golang
+// VerticalPodAutoscalerSpec is the specification of the behavior of the autoscaler.
+type VerticalPodAutoscalerSpec struct {
+  // Denote the recommender to be used
```

Comment:
// Name of the recommender responsible for generating recommendation for this object.